lib/upipe-modules/upipe_audio_merge.c

```diff
@@ -0,0 +1,618 @@
+ *
+ * @param mgr common management structure
+ * @param uprobe structure used to raise events
+ * @param signature signature of the pipe allocator
+ * @param args optional arguments
+ * @return pointer to upipe or NULL in case of allocation error
+ */
+static struct upipe *upipe_audio_merge_sub_alloc(struct upipe_mgr *mgr,
+                                                 struct uprobe *uprobe,
+                                                 uint32_t signature, va_list args)
+{
+    struct upipe *upipe = upipe_audio_merge_sub_alloc_void(mgr,
+                            uprobe, signature, args);
+    if (unlikely(upipe == NULL)) {
+        return NULL;
+    }
+
+    struct upipe_audio_merge_sub *upipe_audio_merge_sub =
+                            upipe_audio_merge_sub_from_upipe(upipe);
+
+    upipe_audio_merge_sub_init_urefcount(upipe);
+    upipe_audio_merge_sub_init_sub(upipe);
+    upipe_audio_merge_sub->uref = NULL;
+    upipe_audio_merge_sub->flow_def = NULL;
+    upipe_throw_ready(upipe);
+    return upipe;
+}
+
+/** @internal @This processes control commands on an input subpipe of an
+ *  audio_merge pipe.
+ *
+ * @param upipe description structure of the pipe
+ * @param command type of command to process
+ * @param args arguments of the command
+ * @return an error code
+ */
+static int upipe_audio_merge_sub_control(struct upipe *upipe,
+                                         int command, va_list args)
+{
+    UBASE_HANDLED_RETURN(upipe_audio_merge_sub_control_super(upipe, command, args));
+    UBASE_HANDLED_RETURN(upipe_control_provide_request(upipe, command, args));
+    switch (command) {
+        case UPIPE_SET_FLOW_DEF: {
+            struct uref *flow_def = va_arg(args, struct uref *);
+            return upipe_audio_merge_sub_set_flow_def(upipe, flow_def);
+        }
+
+        default:
+            return UBASE_ERR_UNHANDLED;
+    }
+}
+
+/** @internal @Copies data from the input urefs to an output buffer.
+ *
+ * @param upipe description structure of the pipe
+ * @param out_data the output buffer
+ */
+static void upipe_audio_merge_copy_to_output(struct upipe *upipe, float **out_data)
+{
+    int8_t out_channel_count = 0;
+    struct upipe_audio_merge *upipe_audio_merge = upipe_audio_merge_from_upipe(upipe);
+    struct uchain *uchain;
+    int i, j;
+
+    ulist_foreach (&upipe_audio_merge->inputs, uchain) {
+        struct upipe_audio_merge_sub *upipe_audio_merge_sub =
+            upipe_audio_merge_sub_from_uchain(uchain);
+
+        uint8_t planes = 0;
+        UBASE_ERROR(upipe, uref_sound_flow_get_planes(upipe_audio_merge_sub->flow_def, &planes));
+
+        float *in_data[planes];
+        UBASE_ERROR(upipe, uref_sound_read_float(upipe_audio_merge_sub->uref, 0, -1, (const float**)in_data, planes));
+
+        uint64_t input_num_samples = 0;
+        UBASE_ERROR(upipe, uref_sound_flow_get_samples(upipe_audio_merge_sub->uref, &input_num_samples));
+
+        uint64_t output_num_samples = 0;
```

Comment:
Anything flow_def related can be obtained outside the loop.

Reply:
moved `output_samples` and `output_num_channels` out of the loop